Lilypad-Tech/lilypad
# feat: Add run job traces

### Review Type Requested (choose one):


### Summary

This pull request implements the following changes:

- [x] Add OpenTelemetry apparatus
  - [x] Add tracer provider
    - [x] Add HTTP exporter
    - [x] Add resource attributes, including service name, version, OS, architecture, GPU info, network (e.g. "testnet"), and address
  - [x] Add `--disable-telemetry` CLI flag
  - [x] Add `telemetry_url` config
    - [x] Set endpoints for local development and production
  - [x] Add `telemetry_token` config 
  - [x] Add `GetOTelServiceName` helper function and table test 
- [x] Add Resource Provider run job trace
  - [x] Add span attributes with information about associated deal 
  - [x] Add span events for run job steps
  - [x] Log deal ID
  - [x] Debug log trace ID on starting the trace
- [x] Disable telemetry in integration tests 
- [x] Move `Version` to system package to make it accessible as a resource attribute 
- [x] Redact private key from logs
- [x] Debug log public address at initialization (all services)
- [x] Debug log Job Offer ID in job creator (to correlate job creator run with resource provider run job trace)
- [x] Upgrade to Go 1.22
- [x] Add debug logs to `test` GitHub action workflow

### Task/Issue reference

Closes: #293 

### Details

This pull request adds our initial tracing implementation with a single trace over resource provider job runs.

We collect the following information about resource providers:

- Operating system
- Architecture
- GPU (if any)
- Wallet address
- Network (`dev`, `devnet`, or `testnet`)

This information will also be collected for any service where we add tracing.

We collect the following information about jobs:

- Deal ID
- Job creator address
- Resource provider address
- Job offer CID
- Resource offer CID
- Module repo
- Module hash (the git hash, also maybe a tag)
- Result ID
- Transaction hash from adding the result on chain

The public address, job offer ID, deal ID, trace ID logs are identifiers for looking up traces.

### How to test this code? 

Run our forthcoming observability implementation (https://github.com/Lilypad-Tech/observability/tree/dev-ronelcanaria/feat-add-alloy):

```sh
./stack observe-up
```

Navigate to the Grafana Dashboard at `http://localhost:3000`. Start the full Lilypad stack, then run a job. 

The resource provider should send the trace to the observability stack where it can be viewed in `Home > Explore`, select "Tempo" at the top, and search. It's possible to filter on `run_job` for the span name, but shouldn't be needed because we only have one trace.
